Topics/Tech_Stacks/Metaprogramming.md

```diff
@@ -0,0 +1,142 @@
+communication between the server and clients.
+
+### JSON Schema
+
+Every class is serialized to JSON in the following manner. 
+
+```json
+{
+    class_name: class_field
+}
+```
+
+for example the class
+
+```python
+class SignInMessage(Message):
+    def __init__(self, username, password):
+        self.username = username
+        self.password = password
+```
+
+would be serialized to
+
+```json
+{
+    "SignInMessage": {
+        "username": "username",
+        "password": "password"
+    }
+}
+```
+
+Note: any other format can be chosen. This is just the one I chose. You only need to modify the 
+to_json_message and from_json_message methods in the Message class.
+
+The magic happens in the message class from which all other message types inherit.
+
+```python
+
+class Message(BaseModel):
+    #this is the method that generates the json serialization.
+    # for example class SignIn(username: str, password: str) would return
+    # {"SignIn": {"username": "username", "password": "password"}}
+    def to_json_message(self) -> str:
+        # we find the name of the class through the __class__ attribute.
+        name = self.__class__.__name__
+        # we convert the class to a dictionary using the model_dump method.
+        data = self.model_dump_json()
+        # then we add the class name as a key to the dictionary. and convert the dictionary to json.
+        return json.dumps({name: data})
+
+    # this is the method that generates the json deserialization and returns a 
+    # specific Message object that we can use
+    # for dynamic dispatch.
+    # for example {"SignIn": {"username": "username", "password": "password"}} 
+    # would return a SignIn object with the username and password fields set.
+    @staticmethod
+    def from_json_message(message: str) -> Message:
+        # we convert the json to a dictionary.
+        dictionary = json.loads(message)
+        # extract the class name and the class fields from the dictionary.
+        name, data = next(iter(dictionary.items()))
+        data = json.loads(data)
+        # we look up the class in the global scope and return a new instance of 
+        # the class with the data as the arguments.
+        cls = globals()[name]
+        return cls(**data)
+
+    # this is the method that handles the logic of the message. We need to implement
+    # this method in all our Message subclasses.
+    # for example class SignIn(username: str, password: str) would return a SignInResponse object.
+    # this is how we get dynamic dispatch to work. We can call the respond method on any Message object and it will return the correct response. as defined by the subclass.
+    def respond(self) -> Message:
+        raise NotImplementedError
+```
+
+Then assuming we have a server object that can send and receive JSON. 
+the server architecture is as follows.
+
+```python
+    while True:
```

Comment:
Mention the importance of handling exceptions, especially in a networked environment as there is non in the for-loop for what I assume was simplicity's sake

Reply:
I added a try block wrap as well a explanation for why it should exist.